src/stackdriver-nozzle/stackdriver/metrics_buffer.go

```diff
@@ -16,12 +16,83 @@
 
 package stackdriver
 
-import "reflect"
+import (
+	"context"
+	"github.com/mitchellh/hashstructure"
+	"reflect"
+	"sync"
+	"time"
+)
 
 type MetricsBuffer interface {
 	PostMetric(*Metric)
 }
 
+type timerMetricsBuffer struct {
+	adapter MetricAdapter
+	errs    chan error
+	size    int
+	ticker  *time.Ticker
+	ctx     context.Context
+
+	metricsMu sync.Mutex // Guard metrics
+	metrics   metricsMap
+}
+
+func NewTimerMetricsBuffer(ctx context.Context, frequency time.Duration,
+	size int, adapter MetricAdapter) (MetricsBuffer, <-chan error) {
+	errs := make(chan error)
+	mb := &timerMetricsBuffer{
+		adapter: adapter,
+		errs:    errs,
+		metrics: make(map[string]*Metric),
+		size:    size,
+		ctx:     ctx,
+		ticker:  time.NewTicker(frequency),
+	}
+	mb.start()
+	return mb, errs
+}
+
+func (mb *timerMetricsBuffer) PostMetric(metric *Metric) {
+	mb.addMetric(metric)
+}
+
+func (mb *timerMetricsBuffer) addMetric(newMetric *Metric) {
+	k, _ := hashstructure.Hash(newMetric.Labels, nil)
+	mb.metricsMu.Lock()
+	defer mb.metricsMu.Unlock()
+	mb.metrics[newMetric.Name+string(k)] = newMetric
```

Comment:
Yep! Worth noting I've confirmed that `hashstructure` consistently hashes labels, even though that's a `map[string]string` with no guaranteed order.

This is a naive take. We could make `metrics` a `map[string][]*Metric`, appending dupes and allowing something downstream to decide how to handle the dupes (average them, take earliest, latest, etc). I haven't thought much about where we'd do that, but probably here or the adapter.